File: test_kubernetes/Java/app/src/main/java/jorge/cardona/kubernetes/usecases/Info.java
package jorge.cardona.kubernetes.usecases;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import jorge.cardona.kubernetes.models.Config;
import lombok.Data;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Configuration;
import org.springframework.core.SpringVersion;
import org.springframework.web.client.RestTemplate;

import java.io.*;
import java.net.InetAddress;
import java.net.UnknownHostException;
import java.nio.charset.Charset;
import java.nio.file.Path;
import java.util.HashMap;
import java.util.Map;

@Configuration
@Data
public class Info {

    @Autowired
    Config config;

    public String setDirectorioDeArchivos() {

        File directorio = new File(config.getDirectorio());
        String mensaje = null;

        if (!directorio.exists()) {
            try {
                directorio.mkdirs();
                mensaje = "Directorio creado";
            } catch (SecurityException e) {
                mensaje = "No se pudo crear el directorio";
            }
            return mensaje;

        } else {
            return "El directorio ya estaba creado";
        }
    }

    public String setContentFile(String message) throws IOException {

        setDirectorioDeArchivos();
        FileWriter archivoDeDatos = new FileWriter(config.getFullpath(), Charset.forName("UTF-8"), true);
        archivoDeDatos.append("Java Version " + System.getProperty("java.version") + " Host " + InetAddress.getLocalHost().getHostAddress() + "  Informacion Adicionada "+ message + "\n");
        archivoDeDatos.close();

        return getContentFile();

    }

    public String getContentFile() throws IOException {

        createFile();

        Path filePath = Path.of(config.getFullpath());

        StringBuilder contentBuilder = new StringBuilder();
        try (BufferedReader br = new BufferedReader(new FileReader(String.valueOf(filePath)))) {
            String sCurrentLine;
            while ((sCurrentLine = br.readLine()) != null) {
                contentBuilder.append(sCurrentLine).append("\n");
            }
        } catch (IOException e) {
            e.printStackTrace();
        }

        return contentBuilder.toString();
    }


    public Map<String, String> getEnvironmentVariables() {
        Map<String, String> envMap = System.getenv();

        return envMap;
    }

    public Map<String, Object> getInternalInfo() throws UnknownHostException {

        Map<String, Object> map = new HashMap<>();
        Map<String, String> environmentVariables = getEnvironmentVariables();

        Runtime.Version runtimeVersion = Runtime.version();
        String version = String.valueOf(runtimeVersion.version().get(0));
        InetAddress hostIp = InetAddress.getLocalHost();
        String ip = InetAddress.getLocalHost().getHostAddress();
        String hostname = InetAddress.getLocalHost().getHostName();

        map.put("Pod - Image  -> Aplication", "Java Version " + System.getProperty("java.version"));
        map.put("Pod - Image  -> Spring Version", SpringVersion.getVersion());
        map.put("Pod - Image  -> Author", "Jorge Cardona");
        map.put("Pod - Image  -> hostname", hostname);
        map.put("Pod - Image  -> ip", ip);
        map.put("Pod - Image  -> hostname/ip", hostIp.toString());

        return map;

    }

    public Map<String, Object> getAllInfo() throws UnknownHostException {

        Map<String, Object> map = getInternalInfo();
        Map<String, String> environmentVariables = getEnvironmentVariables();
        map.put("Pod - Image  -> Environment Variables", environmentVariables);

        return map;

    }

    public Map<Object, Object> getExternalWeb(String url) throws UnknownHostException {

        Map<Object, Object> map = new HashMap();
        Map<String, Object> internalInfo = getAllInfo();

        map.put("Local API Data Information", internalInfo);

        RestTemplate request = new RestTemplate();

        try {
            String response = request.getForObject(url, String.class);

            ObjectMapper mapper = new ObjectMapper();
            JsonNode node = mapper.readTree(response);

            map.put("Response from External API Consumed", node);

        } catch (Exception e) {
            map.put("error", "website not found");
        } finally {
            return map;
        }
    }

    public Map<Object, Object> getExternalApiResponse(String url, int port) throws UnknownHostException, JsonProcessingException {

        Map<Object, Object> map = new HashMap();
        Map<String, Object> internalInfo = getAllInfo();

        map.put("Local API Data Information", internalInfo);

        RestTemplate request = new RestTemplate();
        String address = url + ":" + port;

        try {
            String response = request.getForObject(address, String.class);

            ObjectMapper mapper = new ObjectMapper();
            JsonNode node = mapper.readTree(response);

            map.put("Response from External API Consumed", node);

        } catch (Exception e) {
            map.put("error", "website not found");
        } finally {
            return map;
        }
    }

    public String deleteFile() {

        File file = new File(config.getFullpath());
        String mensaje = null;

        if (file.exists()) {
            try {
                file.delete();
                mensaje = "Archivo eliminado";
            } catch (SecurityException e) {
                mensaje = "No se pudo eliminar el archivo";
            }
            return mensaje;

        } else {
            return "El archivo no existe";
        }
    }

    public String createFile() {

        File file = new File(config.getFullpath());
        String mensaje = null;

        if (!file.exists()) {
            try {
                file.createNewFile();
                mensaje = "Archivo creado";
            } catch (SecurityException | IOException e) {
                mensaje = "No se pudo crear el archivo";
            }
            return mensaje;

        } else {
            return "El archivo ya estaba creado";
        }
    }

}

File: test_kubernetes/Python/app_normal/usecases/info.py
import socket
import sys
from fastapi import __version__
import requests
# poner los dos . puntos literales para importar el modulo, sino genera error al crear la imagen de docker
from models.hosts import Host
import os
from pathlib import Path


directorio = os.getenv('DIRECTORIO', default='/data/files')
archivo = os.getenv('ARCHIVO', default='info.txt')
fullpath = directorio + '/' + archivo

def get_enviroment_variables():
    return os.environ.items()

def get_internal_info():
    hostname=socket.gethostname()
    IPAddr=socket.gethostbyname(hostname)

    return {
		"Pod - Image  -> Aplication": "Python " + sys.version,
        "Pod - Image  -> FastApi Version ": __version__,
        "Pod - Image  -> hostname": socket.gethostname(),
        "Pod - Image  -> Author": "Jorge Cardona",
        "Pod - Image  -> hostname/ip": f"{hostname}/{IPAddr}",
        "Pod - Image  -> ip": IPAddr
    }

def get_complete_info():
    
    actualinfo = get_internal_info()
    actualinfo[ "Pod - Image  -> Environment Variables"] = get_enviroment_variables()
    return actualinfo

def get_external_info(host: Host):
    try:
        response = requests.get(f"{host.hostname}:{host.port}").json()
    except Exception:
        response ={"error": "Host or Port not found"}
    else: 
        response     
    finally:
        return response
        
def get_external_api_response(host: Host):
    
    response = {}
    
    response["Response from External API Consumed"] = get_external_info(host)
    response["Local API Base Data"] = get_complete_info() 
    
    return response
    
def get_external_web(url:str):
    
    try:
        response = requests.get(url).json()
    except Exception:        
        response ={"error": "website not found"}
    else:        
        response["Local API Base Data"] = get_complete_info()        
    finally:
        return response
   
   
def set_directory_files(directory:str=directorio):

    Path(f'{directory}').mkdir(parents=True, exist_ok=True)
    
    os.chdir(directory)
    
    create_file()
    
    return os.listdir()
 
def get_content_file(file_name:str=archivo):
    
    set_directory_files()
    
    try:
        with open(file_name, "r") as file:
            content = file.readlines()
    except Exception:
        content = None
    finally:
        return content
    
    
def save_content_file(file_name:str=archivo, message:str='Informacion Compartida'):
    
    set_directory_files()
    hostname=socket.gethostname()
    
    try:
        with open(file_name, "a") as file:
            file.write(f'Python {sys.version} Host {hostname} Informacion Adicionada-> {message} \n')
    except Exception:
        message ={"error": "File not found"}
    else:
        message = get_content_file()
    finally:
        return message
    
def delete_file(file_name:str=archivo):
    
    try:
        os.remove(archivo)
        os.chdir(directorio)
    except Exception:
        message ={"error": f"El archivo {fullpath} no existe"}
    else:
        message = f"El archivo {fullpath} fue eliminado"
    finally:
        return message
    
def create_file(file_name:str=archivo):
    
    try:
        with open(file_name, "x") as file:
            file.write(f'')
    except Exception:
        message ={"error": f"El archivo {fullpath} ya existe"}
    else:
        message = f"El archivo {fullpath} fue creado"
    finally:
        return message